### PR TITLE
Add CI for automatically building and pushing packages

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

See https://github.com/crossplane/crossplane/pull/4885 for the framework/philosophy behind this implementation.

Closes https://github.com/crossplane/function-template-go/pull/16

This PR was heavily influenced by #16 (thanks @pedjak). The main differences are:

* Avoid introducing a Makefile.
* Don't test and lint inside the multi-stage build

These differences both come from a goal to keep build and CI as simple and lightweight as possible:

* I don't think the learning curve of Make is worth the value it brings (`make test` doesn't add much vs `go test`)
* I think realistically folks are going to need the Go toolchain installed to develop a Function so running `go test` inside Docker is worth the price of a more complicated / hard to follow Dockerfile.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] ~Added or updated unit tests for my change.~

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  push:
+    branches:
+      - main
+      - release-*
+  pull_request: {}
+  workflow_dispatch:
+    inputs:
+      version:
+        description: Package version (e.g. v0.1.0)
+        required: false
```
Comment (negz) [inline .github/workflows/ci.yml:13]: I'm not sure everyone will want to take on the hassle of maintaining release branches, but they'll presumably at least want to tag and push versioned releases of their Function. So the goal here is:

* When the CI workflow runs against a PR or the main branch we build and push it as `v0.0.0-gitsha`.
* You can run the CI workflow manually and provide whatever version you want.

I don't love that the `v0.0.0-gitsha` pattern loses semver context, but I don't really want to inflict assumptions around how git tags are used on folks.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  GO_VERSION: '1.21.3'
+  GOLANGCI_VERSION: 'v1.54.2'
+  DOCKER_BUILDX_VERSION: 'v0.11.2'
+  XP_CHANNEL: master   # TODO(negz): Pin to stable once v1.14 is released.
+  XP_VERSION: current  # TODO(negz): Pin to a version once v1.14 is released.
+
+  # Common users. We can't run a step 'if secrets.USR != ""' but we can run
+  # a step 'if env.USR' != ""', so we copy these to succinctly test whether
+  # credentials have been provided before trying to run steps that need them.
+  XPKG_ACCESS_ID: ${{ secrets.XPKG_ACCESS_ID }}
```
Comment (negz) [inline .github/workflows/ci.yml:26]: Nothing will be pushed to xpkg.upbound.io unless this secret (and the corresponding token secret) are created.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          target: image
+          build-args:
+            GO_VERSION=${{ env.GO_VERSION }}
+          outputs: type=docker,dest=runtime-${{ matrix.arch }}.tar
+      
+      - name: Setup the Crossplane CLI
+        run: "curl -sL https://raw.githubusercontent.com/crossplane/crossplane/master/install.sh | sh"
+
+      - name: Build Package
+        run: ./crossplane xpkg build --package-file=${{ matrix.arch }}.xpkg --package-root=package/ --embed-runtime-image-tarball=runtime-${{ matrix.arch }}.tar
```
Comment (negz) [inline .github/workflows/ci.yml:104]: http://github.com/crossplane-contrib/xpkg-action/ is currently broken because it was never updated to accomodate https://github.com/crossplane/crossplane/issues/3345.

We may want to revamp it and use it in future. I'm on the fence though. It feels more polished/professional to use an Action, but honestly just `run: ./crossplane` seems to work pretty well.

File: Dockerfile [hunk 4]
```diff
@@ -1,37 +1,30 @@
-FROM golang:1.21 as build-stage
+# syntax=docker/dockerfile:1
```
Comment (negz) [inline Dockerfile:1]: I think this is a fairly minimal/idiomatic modern Dockerfile, but some of the features like `--mount=type=cache` for caching and using `--platform=${BUILDPLATFORM}` for cross-compiling were new to me. The Docker documentation covers these all well, and this is pretty close to their multi-stage and multi-platform examples for Go but it may still be worth adding a bit more commentary to this file.

**Comment by pedjak** [comment]
> Don't test and lint inside the multi-stage build

What speak against having them in the multi-stage build @negz ?

**Comment by negz** [comment]
@pedjak I think folks are going to need the Go toolchain installed to develop a Function so running go test inside Docker is worth the price of a more complicated / hard to follow Dockerfile.

**Comment by negz** [comment]
I was leaving this in draft because I planned to update `README.md`. However on reflection I think I'd like the readme for this template (and others) to be very small, and move the bulk of the "how to write a Function" into the docs.

I'll just leave the README for now, and replace it in a future PR once there's a docs page to link to.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jbw976) [reply .github/workflows/ci.yml:13]: Yeah i'm a bit conflicted on this one too - I think the intent to not require release branches is reasonable, but I wonder if we can push folks towards using tags at least, so we can get a meaningful version number always.

The build submodule uses tags if the exist and if they don't then it falls back to `v0.0.0`:
https://github.com/upbound/build/blob/master/makelib/common.mk#L214-L225

Is that type of conditional logic possible here?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        description: Package version (e.g. v0.1.0)
+        required: false
+
+env:
+  # Common versions
+  GO_VERSION: '1.21.3'
+  GOLANGCI_VERSION: 'v1.54.2'
+  DOCKER_BUILDX_VERSION: 'v0.11.2'
+  XP_CHANNEL: master   # TODO(negz): Pin to stable once v1.14 is released.
+  XP_VERSION: current  # TODO(negz): Pin to a version once v1.14 is released.
```
Comment (jbw976) [inline .github/workflows/ci.yml:21]: where are these actually used? I couldn't find them just with ctrl+F 🤔

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+  # Common users. We can't run a step 'if secrets.USR != ""' but we can run
```
Comment (jbw976) [inline .github/workflows/ci.yml:23]: i'm not familiar with `secrets.USR` and `env.USR` are defined and how they relate to `XPKG_ACCESS_ID` - are they common github vars? any extra context that should be added here?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+          platforms: linux/${{ matrix.arch }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          target: image
+          build-args:
+            GO_VERSION=${{ env.GO_VERSION }}
+          outputs: type=docker,dest=runtime-${{ matrix.arch }}.tar
+      
+      - name: Setup the Crossplane CLI
+        run: "curl -sL https://raw.githubusercontent.com/crossplane/crossplane/master/install.sh | sh"
```
Comment (jbw976) [inline .github/workflows/ci.yml:101]: should this stay `master` after v1.14 ships?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+      - name: Build Package
+        run: ./crossplane xpkg build --package-file=${{ matrix.arch }}.xpkg --package-root=package/ --embed-runtime-image-tarball=runtime-${{ matrix.arch }}.tar
+      
+      - name: Upload Single-Platform Package
```
Comment (jbw976) [inline .github/workflows/ci.yml:106]: how is this upload single platform package step different than `push` step below?

edit: ah, this is to upload as an artifact on GitHub, but `push` below uploads the package to marketplace

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        uses: actions/checkout@v4
+
+      - name: Download Single-Platform Packages
+        uses: actions/download-artifact@v3
+        with:
+          name: packages
+          path: .
+
+      - name: Setup the Crossplane CLI
+        run: "curl -sL https://raw.githubusercontent.com/crossplane/crossplane/master/install.sh | sh"
```
Comment (jbw976) [inline .github/workflows/ci.yml:129]: same question about if they should still use `master` after v1.14 ships

File: Dockerfile (hunk 4 above)
Comment (jbw976) [reply Dockerfile:1]: +1 on more commentary 🤓 - i can guess at what those args do, but understanding some nuance of how they interact with the GH actions environment, if anything would need to change for a local dev environment, etc. would be helpful

**Review by jbw976 — COMMENTED** [review]
awesome @negz, this is a great step for streamlining the building of functions and making the development process approachable 💪

**Comment by pedjak** [comment]
> I think folks are going to need the Go toolchain installed to develop a Function so running go test inside Docker is worth the price of a more complicated / hard to follow Dockerfile.

Adding a few commentary would help to demystify the new changes and how to use them. Ofcourse, people can and should still use their local toolchain, but having everything in Dockerfile would greatly simplify the CI job. In addition to that, developers would get a possibility to verify the code locally against the toolchain used by CI, before even creating a PR.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (pedjak) [inline .github/workflows/ci.yml:104]: are we able to build multi-arch package?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Build Runtime
+        id: image
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          platforms: linux/${{ matrix.arch }}
```
Comment (pedjak) [inline .github/workflows/ci.yml:92]: Instead of having a build per arch, we could actually build all images at once, saving time.

File: Dockerfile [hunk 11]
```diff
@@ -1,37 +1,30 @@
...
+    GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -o /function .
+
+# Produce the Function image.
+FROM gcr.io/distroless/base-debian11 AS image
 WORKDIR /
-
-COPY --from=build-stage /function /function
-COPY --from=package-stage /package.yaml /package.yaml
-
+COPY --from=build /function /function
```
Comment (pedjak) [inline Dockerfile:27]: the image does not need to contain `crossplane.yaml` anymore?

Comment (negz) [reply Dockerfile:27]: The `crossplane xpkg build` command takes care of this now. It does it "better" in that it compiles `package.yaml` and puts it in a layer with a special annotation. That way the package manager can pull just the layer it needs to read `package.yaml`, not the whole (somewhat large) image. 

See https://github.com/crossplane/crossplane/blob/master/contributing/specifications/xpkg.md for details.

File: Dockerfile [hunk 12]
```diff
@@ -1,37 +1,30 @@
...
-
-# TODO(negz): Use a proper Crossplane package building tool. We're abusing the
-# fact that this image won't have an io.crossplane.pkg: base annotation. This
-# means Crossplane package manager will pull this entire ~100MB image, which
-# also happens to contain a valid Function runtime.
-# https://github.com/crossplane/crossplane/blob/v1.13.2/contributing/specifications/xpkg.md
-WORKDIR /package
-COPY package/ ./
+# We use the latest Go 1.x version unless asked to use something else.
+ARG GO_VERSION=1
```
[the remark this answers is not in the document]
Comment (negz) [reply Dockerfile:4]: This is just a sane default value. The CI job overrides it with its value - see `build-args` of the "Build Runtime" job.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (negz) [reply .github/workflows/ci.yml]: It was meant as an example value to demonstrate the different between secrets.WHATEVER and env.WHATEVER. I deleted this comment.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (negz) [reply .github/workflows/ci.yml:24]: Added a comment to clarify.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (negz) [reply .github/workflows/ci.yml:113]: I think so, yes. The XP_CHANNEL and XP_VERSION tell the script (from master) to download a specific version.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (negz) [reply .github/workflows/ci.yml:118]: Added a comment to the top of this job to explain what is happening.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (negz) [reply .github/workflows/ci.yml:144]: Answered above. 🙂

File: .github/workflows/ci.yml (hunk 2 above)
Comment (negz) [reply .github/workflows/ci.yml:29]: Added (another) comment to clarify.

File: Dockerfile (hunk 4 above)
Comment (negz) [reply Dockerfile:1]: Added comments to clarify.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (negz) [reply .github/workflows/ci.yml:104]: This matrix job uses one GitHub runner per architecture to build in parallel for each architectures. I haven't tested but I presume this is faster than having one GitHub runner building all architectures (even if that one runner builds in parallel).

https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners

File: .github/workflows/ci.yml (hunk 3 above)
Comment (negz) [reply .github/workflows/ci.yml:116]: A multi-platform package (or OCI image) is actually N single-platform OCI images with an "OCI index" abstraction on top of them. The index is pretty much an alias from a tag/reference to a set of images. I don't think an xpkg can accept an OCI index as its "runtime" - it wants real image layers that it can embed and add a layer to.

**Comment by negz** [comment]
I feel a need to have something in place here to document for launch, so I'm moving forward with this as is. Thanks for the feedback @pedjak and @jbw976!

File: Dockerfile (hunk 12 above)
[the remark this answers is not in the document]
Comment (pedjak) [reply Dockerfile:4]: Does this mean that `Dockerfile` is only used by CI, and should not be used by developers for local builds?
